Move convergence criteria out of linear solver

nilsvu opened the pull request:
## Proposed changes

This PR moves the convergence criteria out of the linear solver so they may be used also by the nonlinear solver. In the future I can also make them generic to support other iterative algorithms, e.g. FastFlow.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu commented:
Note to reviewers: This literally just moves `LinearSolver/Convergence.{hpp,cpp}` to `Convergence/{Reason,Criteria,HasConverged}.{hpp,cpp}`.

nilsvu commented:
@nilsdeppe @kidder could you review or find reviewers for this?

nilsdeppe commented on src/NumericalAlgorithms/Convergence/HasConverged.hpp:
```diff
@@ -0,0 +1,95 @@
+namespace Convergence {
+
+/*!
+ * \brief Determine whether the \p criteria are met.
```
It's not clear to me this function should be public vs having people use the `HasConverged` class. @nilsleiffischer do you have any reasons to prefer it being public?

nilsvu replied:
discussed this offline and decided to keep as-is

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitor.hpp:
```diff
@@ -96,13 +96,14 @@ struct InitializeResidualMonitor {
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/,
+                    ::Verbosity verbosity,
+                    Convergence::Criteria convergence_criteria) noexcept {
```
`const Convergence::`?

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitor.hpp:
```diff
@@ -96,13 +96,14 @@ struct InitializeResidualMonitor {
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/,
+                    ::Verbosity verbosity,
```
`const ::Verbosity`?

kidder commented:
@moxcodes can you review this? it should be super quick (i.e. a rename of reviewed code)

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Convergence/Test_Criteria.cpp:
```diff
@@ -0,0 +1,26 @@
+  CHECK(criteria != Convergence::Criteria{3, 0., 0.5});
+  CHECK(criteria != Convergence::Criteria{2, 1., 0.5});
+  CHECK(criteria != Convergence::Criteria{2, 0., 0.6});
+  test_serialization(criteria);
```
Could you set the relative residual to not 0. when testing serialization, etc.?

nilsdeppe reviewed (changes requested):
I'm happy for you to squash the changes

moxcodes commented:
Sure, I can review this.

nilsvu committed:
Move convergence criteria out of the linear solver

These may be used also by the nonlinear solver, as well as
other iterative algorithms

nilsvu commented:
Rebased on #1408 and addressed the comments by @nilsdeppe

moxcodes commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:
```diff
@@ -171,7 +173,9 @@ struct UpdateResidual {
                  static_cast<int>(get<::Tags::Verbosity>(box)) >=
                      static_cast<int>(::Verbosity::Quiet))) {
-      Parallel::printf("%s", has_converged);
+      Parallel::printf("The linear solver has converged in %d iterations: %s",
```
should this have format string `%llu` or similar for the size_t ?

nilsvu replied:
Changed to `%zu` now. I don't know exactly what the best-practice here is, so if you do please tell me.

kidder replied:
yes, `%zu` is what I use

moxcodes commented on src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp:
```diff
@@ -236,7 +238,9 @@ struct StoreFinalOrthogonalization {
                  static_cast<int>(get<::Tags::Verbosity>(box)) >=
                      static_cast<int>(::Verbosity::Quiet))) {
-      Parallel::printf("%s", has_converged);
+      Parallel::printf("The linear solver has converged in %d iterations: %s",
```
same question about the format string

kidder commented:
@nilsleiffischer this isn't compiling...

moxcodes commented on src/NumericalAlgorithms/Convergence/HasConverged.hpp:
```diff
@@ -0,0 +1,93 @@
+ * \brief Determine whether the \p criteria are met.
+ *
+ * \note This function assumes the \p iteration_id is that of the next, but
+ * not yet performed step. For instance, a `MaxIteration` criterium of 1 will
```
as a nitpick, I think `criterion` is preferred.

moxcodes reviewed (commented):
I'm happy for you to squash any of these minor suggestions you choose to implement.

nilsvu commented:
Rebased and fixed iwyu. This should be ready @moxcodes @nilsdeppe @sxs-collaboration/spectre-core-devs
